NitroxServer/GameLogic/Entities/Spawning/BatchEntitySpawner.cs

```diff
@@ -312,33 +318,42 @@ private IEnumerable<Entity> CreateEntityWithChildren(EntitySpawnPoint entitySpaw
 
         entityBootstrapperManager.PrepareEntityIfRequired(ref spawnedEntity, deterministicBatchGenerator);
 
-        yield return spawnedEntity;
+        // Order of addition is important, first, the created entity, then its children and the cell root children it created
+        // so that this output can always be used to get the created entity
+        List<Entity> spawnedEntities = [spawnedEntity];
 
-        if (parentEntity == null) // Ensures children are only returned at the top level
+        for (int i = spawnedEntity.ChildEntities.Count - 1; i >= 0; i--)
         {
-            // Children are yielded as well so they can be indexed at the top level (for use by simulation
-            // ownership and various other consumers).  The parent should always be yielded before the children
-            foreach (Entity childEntity in AllChildren(spawnedEntity))
+            Entity spawnedChildEntity = spawnedEntity.ChildEntities[i];
+            // Regular case in which a spawned child would actually need to be spawned under the cell root instead of the current entity
+            if (spawnedChildEntity.ParentId != spawnedEntity.Id)
             {
-                yield return childEntity;
+                spawnedEntity.ChildEntities.RemoveAt(i);
+                spawnedEntities.Add(spawnedChildEntity);
             }
         }
+
+        // Ensures children are only returned at the top level
+        if (parentEntity == null)
+        {
+            // Children are added as well so they can be indexed at the top level
+            // (for use by simulation ownership and various other consumers).
+            spawnedEntities.AddRange(AllChildren(spawnedEntity));
+        }
+
+        return spawnedEntities;
     }
 
-    private IEnumerable<Entity> AllChildren(Entity entity)
+    private static List<Entity> AllChildren(Entity entity)
     {
-        foreach (Entity child in entity.ChildEntities)
-        {
-            yield return child;
+        List<Entity> allChildren = [.. entity.ChildEntities];
 
-            if (child.ChildEntities.Count > 0)
-            {
-                foreach (Entity childOfChild in AllChildren(child))
-                {
-                    yield return childOfChild;
-                }
-            }
+        for (int i = 0; i < allChildren.Count; i++)
```

Comment:
If this is iterating over all children would a for each not be better than getting the index and using that? Less chance of index expectation problems